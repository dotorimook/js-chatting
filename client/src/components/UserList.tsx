import React, { FC } from 'react';
import IUser from 'interfaces/IUser';
import IComponentProps from 'interfaces/IComponentProps';
import styled from 'styled-components';
import Button from './form/Button';
import theme from 'style/theme';
import { applyFlexbox } from 'style/mixins';

interface IUserListProps extends IComponentProps {
  users: IUser[];
  selectable: boolean;
  onSelect:(userId:number) => void;
}

const UserListFrame = styled.ul`
`;

const UserListItem = styled.li`
  height:2em;
  width:100%;

  .list-item-user {
    width:100%;
    height:100%;
    display:block;
    padding:0 1em;
    ${applyFlexbox('row', 'flex-start', 'center')};
    button {
      display:none;
    }
  }
  .list-item-user:hover {
    background-color:${theme.color2};
  }
`;

const UserList:FC<IUserListProps> = (props:IUserListProps) => {
  const {users, selectable, onSelect} = props;
  return (
    <UserListFrame>
      {
        users.map((user:IUser) => (
          <UserListItem key={`user-${user.id}`}>
            <label className='list-item-user'>
              <span className='name'>{user.name}</span>
              {
                selectable &&
                <Button
                  onClick={()=>{onSelect(user.id);}}
                >
                </Button>
              }
            </label>
          </UserListItem>
        ))
      }
    </UserListFrame>
  )
}

export default UserList;