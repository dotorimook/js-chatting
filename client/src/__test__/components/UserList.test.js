import React from 'react';
import 'setupTest';
import {mount} from 'enzyme';
import UserList from 'components/UserList';

describe('UserList Test', ()=>{
  const users = [{
    id: 0,
    username:'test',
    name:'test',
  }];
  const selectable = true;
  const onSelect = jest.fn();
  let ui = mount(<UserList users={users} selectable={selectable} onSelect={onSelect}/>);

  it('matches snapshot', () => {
    expect(ui).toMatchSnapshot();
  });
  
  it('list item number same as user length', () => {
    expect(ui.find('li').length).toBe(users.length);
  });
});