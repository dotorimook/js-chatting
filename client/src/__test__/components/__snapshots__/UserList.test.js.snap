// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Test matches snapshot 1`] = `
<UserList
  onSelect={[MockFunction]}
  selectable={true}
  users={
    Array [
      Object {
        "id": 0,
        "name": "test",
        "username": "test",
      },
    ]
  }
>
  <styled.ul>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "lastClassName": "bFHdpU",
            "rules": Array [
              "
",
            ],
          },
          "displayName": "styled.ul",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "ul",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <ul
        className="sc-bwzfXH bFHdpU"
      >
        <styled.li
          key="user-0"
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": true,
                  "lastClassName": "hSlyzF",
                  "rules": Array [
                    "
  height:2em;
  width:100%;

  .list-item-user {
    width:100%;
    height:100%;
    display:block;
    padding:0 1em;
    ",
                    "
  
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  
  -webkit-flex-wrap: nowrap;
      -moz-flex-wrap: nowrap;
      -ms-flex-wrap: nowrap;
          flex-wrap: nowrap;

  
  -webkit-flex-direction: row;
      -moz-flex-direction: row;
      -ms-flex-direction: row;
          flex-direction: row;

  
  -webkit-justify-content: flex-start;
      -moz-justify-content: flex-start;
      -ms-justify-content: flex-start;
          justify-content: flex-start;
            -ms-flex-pack: flex-start;

  
  -webkit-align-items: center;
      -moz-align-items: center;
      -ms-align-items: center;
          align-items: center;

",
                    ";
    button {
      display:none;
    }
  }
  .list-item-user:hover {
    background-color:",
                    "rgba(250, 243, 221, 1)",
                    ";
  }
",
                  ],
                },
                "displayName": "styled.li",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "li",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <li
              className="sc-htpNat hSlyzF"
            >
              <label
                className="list-item-user"
              >
                <span
                  className="name"
                >
                  test
                </span>
                <Button
                  onClick={[Function]}
                >
                  <styled.button
                    onClick={[Function]}
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bdVaJa",
                            "isStatic": true,
                            "lastClassName": "cCNCGa",
                            "rules": Array [
                              "
  width: 100%;
  font-size:1em;
  border:none;
  padding: 1em 1em;
  background-color:transparent;

  .material-icons {
    vertical-align:bottom;
    &.left {
      margin-right:.2em;
    }
  }
  
  transition: all 0.2s ease;
  &:hover {
    background-color:",
                              "rgba(184, 242, 230, 1)",
                              "
  }
  &:active {
    transform:translateY(2px);
  }
  
  &[type=submit] {
    background-color:",
                              "rgba(184, 242, 230, 1)",
                              ";
    &:hover {
      background-color:",
                              "rgba(250, 243, 221, 1)",
                              ";
    }
  }
",
                            ],
                          },
                          "displayName": "styled.button",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bdVaJa",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                    >
                      <button
                        className="sc-bdVaJa cCNCGa"
                        onClick={[Function]}
                      />
                    </StyledComponent>
                  </styled.button>
                </Button>
              </label>
            </li>
          </StyledComponent>
        </styled.li>
      </ul>
    </StyledComponent>
  </styled.ul>
</UserList>
`;
